packages/rx-utils/src/context2.utils.ts

```diff
@@ -0,0 +1,61 @@
+import { getReaderM, ReaderM1 } from 'fp-ts/lib/ReaderT';
+import { Monad2 } from 'fp-ts/lib/Monad';
+import { map as readerMap, Reader } from 'fp-ts/lib/Reader';
+import { sequenceT } from 'fp-ts/lib/Apply';
+import { Omit } from 'typelevel-ts';
+import { combineLatest, Observable } from 'rxjs';
+import {
+	ProductLeft,
+	ProductMap,
+	ProjectMany,
+} from '@devexperts/utils/src/typeclasses/product-left-coproduct-left/product-left-coproduct-left.utils';
+import { deferReader } from '@devexperts/utils/src/adt/reader.utils';
+import { pipe, pipeable } from 'fp-ts/lib/pipeable';
+import { memoOnce } from '@devexperts/utils/src/function/memoize';
+import { eqShallow } from '@devexperts/utils/src/typeclasses/eq/eq.utils';
+import { NonEmptyArray } from 'fp-ts/lib/NonEmptyArray';
+import { array } from 'fp-ts/lib/Array';
+import { URI as URIObservable } from 'fp-ts-rxjs/lib/Observable';
+import { observable } from './observable.utils';
+
+export const URI = '@devexperts/dx-utils//Context';
+export type URI = typeof URI;
+declare module 'fp-ts/lib/HKT' {
+	interface URItoKind2<E, A> {
+		[URI]: Context<E, A>;
+	}
+}
+
+export interface Context<E, A> extends Reader<E, Observable<A>> {}
+
+const memo = memoOnce(eqShallow);
+export const contextInstance: Monad2<URI> & ProductLeft<URI> & ReaderM1<URIObservable> = {
+	URI,
+	...getReaderM(observable),
+	asks: f => memo(e => observable.of(f(e))),
```

Comment:
Could you explain, why do we need `memo` here? I can't come up with cases where we would call `asks` several times sequentially with the same argument.

Reply:
The most obvious:
```typescript
interface Env {
  a: number
}
const a = context.asks((e: E) => 'some heavy computation' + e.a + new Date()))
a({ a: 1, b: Math.random() })
a({ a: 1, b: Math.random() })
a({ a: 1, b: Math.random() })
a({ a: 1, b: Math.random() })
a({ a: 1, b: Math.random() })
```

Each context call will rerun the inner computation although dependencies didn't change. This leads to a massive performance regression. However if we tweak `asks` and `context.combine` with shallow memo this won't happen.

Reply:
>  I can't come up with cases where we would call asks several times sequentially with the same argument.

Just pass context to combine several times (like in several parts of application you render the same container - combine the same context):
```typescript
declare const A: Context<{ a: number }, ReactComponent>
const A1 = context.combine(A, (A) => <div><A/></div>)
const A2 = context.combine(A, (A) => <div><A/></div>)
const B = context.combine(A1, A2, (A1, A2) => <div><A1/><A2/></div>)
```